### PR TITLE
Pipe Properties Improvements

### PR DESCRIPTION
**Pull request by TechLord22**
**What:**
This PR makes improvements and additions to Fluid Pipes.

Fluid pipes now have 4 core abilities: supporting gases, acids, cryogenics, and plasmas. A fluid pipe can have any, none, or all of these abilities. This will mean that most pipes will not be good for all purposes, which will lead to players using a more diverse set of pipes for automation.

Fluid pipe stats have been rebalanced for all fluid pipes, such that the speeds of pipes are more balanced to the tier they are unlocked at. A player can attain very high throughput early on, but it may require huge pipes, while that same throughput can be much more easily achieved with better smaller sized pipes. Wrought Iron and Invar pipes have been removed as a result.

Fluid pipe damage has also been redone. There are now damage maximums when a player touches a hot or cold pipe, so they should not be instantly killed with extreme temperatures unless they are low health.

A bug in checking whether fluid pipes can hold a given fluid has also been fixed. Previously, pipes would only check if they are able to contain a fluid when it is transferred, causing some fluids to be able to sit in pipes when there is nowhere to send it, despite the pipe not being able to carry it.

When a pipe is unable to carry a specific fluid, different things happen depending on the fluid contained. All pipes will play a sound to alert the player of their mistake, and will also emit particles depending on what is inside.

* **Gasous Fluids**: 10% of the contained fluid will be voided. It will then apply double the regular contact temperature damage (if the temperature will hurt in the first place) to entities around the pipe, simulating the leaking of hot or cold gas. There is also a 1/7 chance for the pipe to explode (only destroys the pipe), and a 1/3 chance when the fluid is too hot.
* **Acidic Fluids**: 25% of the contained fluid will be voided. 2 chemical damage is applied to entities in a smaller area around the pipe, which also will inflict poison. 10% chance to destroy the pipe. 
* **Hot or Plasma Fluids**: 75% of the contained fluid will be voided. 25% chance for the area surrounding the pipe will be lit on fire. Entities around the pipe get double the contact temperature damage applied, if the contained fluid is plasma. 10% chance to turn the pipe into fire.
* **Cryogenic Fluids**: 75% of the contained fluid will be voided. Entities around the pipe receive double the contact temperature damage. 10% chance for the pipe to be destroyed.

New Damage Types have been added and existing ones modified to further expand upon what happens when in contact with pipes.

* **Heat Damage**: Also counts as fire damage. Entities naturally resistant to fire and entities with fire resistance are not effected.
* **Frost Damage**: Entities naturally resistant to the cold and entities with frost walker boosts are not effected. Frost Walker boots do however take damage.
* **Chemical Damage**: Entities that cannot breathe chemicals are resistant, such as skeletons. It applies regular damage and also poisons for a random amount of time.

Additionally, plasma supporting pipes will hold any plasma regardless of temperature, but will respect temperature for non-plasmas. Furthermore, this PR fixes a bug where plasma state fluid tooltips were not applied correctly.

**Outcome:**
Overhauled Pipe Properties to be more complex. Improved entity damaging with fluids in pipes. Rebalanced pipe stats.

**Possible compatibility issue:**
Players using existing pipes will most likely need to break and replace them in order for proper fluid containment rules to function again, only for fluids needing new handling as of this pr (such as acids and plasmas). 

Players previously using Wrought Iron or Invar pipes will still have their pipes functioning in world, though the pipe stats will be different from the original values, which may cause pipe burning or other problems with existing automation. Players will be unable to recycle these pipes after updating, since no recipes will be generated for them. They should do so before updating.

### CODE REVIEW

File: src/main/java/gregtech/common/pipelike/fluidpipe/BlockFluidPipe.java [hunk 1]
```diff
@@ -110,36 +110,21 @@ public boolean isHoldingPipe(EntityPlayer player) {
...
-            List<Integer> temps = new ArrayList<>();
+
+            // apply temperature damage for each pipe
             for (FluidTank tank : ((TileEntityFluidPipeTickable) pipe).getFluidTanks()) {
                 if (tank.getFluid() != null && tank.getFluid().amount > 0) {
-                    temps.add(tank.getFluid().getFluid().getTemperature(tank.getFluid()));
+                    int temperature = tank.getFluid().getFluid().getTemperature(tank.getFluid());
+                    EntityDamageUtil.applyTemperatureDamage(entityLiving, temperature, 1.0F, 5);
                 }
             }
```
Comment (brachy84) [inline src/main/java/gregtech/common/pipelike/fluidpipe/BlockFluidPipe.java:127]: Adding the temperatures of all fluids as damage seems brutal. I used the average temperature before.

**Review by brachy84 — APPROVED** [review]
Looks good overall. I noticed that you inline `if()` a lot. I personally think we should always use the full
```java
if(true) {
}
```

File: src/main/java/gregtech/common/pipelike/fluidpipe/BlockFluidPipe.java [hunk 2]
```diff
@@ -110,35 +110,25 @@ public boolean isHoldingPipe(EntityPlayer player) {
...
                 return;
-            List<Integer> temps = new ArrayList<>();
+
+            // apply temperature damage for each pipe
+            int temperature = 0;
+            int nonEmptyTanks = 0;
             for (FluidTank tank : ((TileEntityFluidPipeTickable) pipe).getFluidTanks()) {
                 if (tank.getFluid() != null && tank.getFluid().amount > 0) {
-                    temps.add(tank.getFluid().getFluid().getTemperature(tank.getFluid()));
+                    temperature += tank.getFluid().getFluid().getTemperature(tank.getFluid());
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/pipelike/fluidpipe/BlockFluidPipe.java:126]: Couldn't this average make it so that neither Cyrogenic nor Heat damage were applied if the temperature values were just right? It might be better to sum all the pipes that would apply heat damage and all the pipes that would apply cold damage in separate variables, and then make two calls to `applyTemperatureDamage`, one for each variable.

File: src/main/java/gregtech/loaders/recipe/handlers/PipeRecipeHandler.java [hunk 3]
```diff
@@ -82,11 +83,11 @@ private static void processPipeTiny(OrePrefix pipePrefix, Material material, IMa
                     .duration((int) (material.getMass()))
                     .EUt(6 * getVoltageMultiplier(material))
                     .buildAndRegister();
+        } else {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/loaders/recipe/handlers/PipeRecipeHandler.java:86]: Don't really agree with this change. The hand crafting recipes should always be available in my opinion.

File: src/main/java/gregtech/loaders/recipe/handlers/PipeRecipeHandler.java [hunk 4]
```diff
@@ -51,7 +52,7 @@ private static void processRestrictivePipe(OrePrefix pipePrefix, Material materi
         else if (pipePrefix == OrePrefix.pipeHugeRestrictive) unrestrictive = OrePrefix.pipeHugeItem;
         else return;
 
-        RecipeMaps.ASSEMBLER_RECIPES.recipeBuilder()
+        RecipeMaps.PACKER_RECIPES.recipeBuilder()
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/loaders/recipe/handlers/PipeRecipeHandler.java:55]: Why the change to use the Packager? It does not really make sense to use it here.

**Review by ALongStringOfNumbers — COMMENTED** [review]
For the Wooden and Treated Wood pipes, since they cannot handle any of the special cases, including the case of gases, when showing the fluid containment info tooltip, could there be some mention that these things are not accepted, so holding shift does not just show an empty tooltip, as it does now.

An idea for this is to always show all the possible containment options, instead of just the applicable ones.
EG:
(Green) Can Handle Gases
(Green) Can Handle Acids
(Red)     Cannot Handle Cyrogenics
(Red)     Cannot Handle All Plasmas

Showing all the containment information in the tooltip would be fine in my opinion, as the information is hidden behind holding shift and not always in the players face.

----

In my opinion there needs to be more pipes that can handle Cyrogenic materials. Right now there is only Neutronium, Naquadah, Duranium, and Stainless Steel. Which is three end game materials, and one midish game material.

I would suggest something like Niobium-Titanium, Iridium, or maybe PBI as an in between the starting pipe of stainless steel and the really end game materials.

----

File: src/main/java/gregtech/common/pipelike/fluidpipe/BlockFluidPipe.java [hunk 5]
```diff
@@ -110,35 +110,26 @@ public boolean isHoldingPipe(EntityPlayer player) {
...
+                    if (tank.getFluid() != null && tank.getFluid().amount > 0) {
+                        maxTemperature = Math.max(maxTemperature, tank.getFluid().getFluid().getTemperature(tank.getFluid()));
+                        minTemperature = Math.min(minTemperature, tank.getFluid().getFluid().getTemperature(tank.getFluid()));
+                    }
+                }
+                if (maxTemperature != Integer.MIN_VALUE) {
+                    EntityDamageUtil.applyTemperatureDamage((EntityLivingBase) entityIn, maxTemperature, 1.0F, 5);
+                }
+                if (minTemperature != Integer.MAX_VALUE) {
+                    EntityDamageUtil.applyTemperatureDamage((EntityLivingBase) entityIn, minTemperature, 1.0F, 5);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/pipelike/fluidpipe/BlockFluidPipe.java:131]: This will be an issue if there is only 1 fluid in the pipe. The max temperature and the min temperature will be set to the same thing, and so `applyTemperatureDamage` will be called twice, because `maxTemperature` and `minTemperature` are both not equal to the Max/Min values.
